Data/models/RecentSearches.swift

```diff
@@ -0,0 +1,123 @@
+    @NSManaged public var searchType: Int32
+    @NSManaged public var text: String?
+    @NSManaged public var websiteUrl: String?
+    @NSManaged public var dateAdded: Date?
+    @NSManaged public var order: Int32
+    
+    public class func frc() -> NSFetchedResultsController<RecentSearch> {
+        let context = DataController.viewContext
+        let fetchRequest = NSFetchRequest<RecentSearch>()
+        fetchRequest.entity = RecentSearch.entity(context)
+        fetchRequest.fetchBatchSize = 5
+        
+        let orderSort = NSSortDescriptor(key: "order", ascending: true)
+        let createdSort = NSSortDescriptor(key: "dateAdded", ascending: false)
+        fetchRequest.sortDescriptors = [orderSort, createdSort]
+        
+        return NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: context,
+                                          sectionNameKeyPath: nil, cacheName: nil)
+    }
+    
+    public static func addItem(type: RecentSearchType, text: String?, websiteUrl: String?) {
+        DataController.perform(context: .new(inMemory: false), save: false) { context in
+            let item = RecentSearch(context: context)
+            item.searchType = type.rawValue
+            item.text = text
+            item.websiteUrl = websiteUrl
+            item.dateAdded = Date()
+            item.order = -9999
+            
+            RecentSearch.saveContext(context)
+            RecentSearch.reorderItems(context: context)
+            RecentSearch.saveContext(context)
+        }
+    }
+    
+    public static func getItem(text: String) -> RecentSearch? {
+        return RecentSearch.first(where: NSPredicate(format: "text == %@", text))
+    }
+    
+    public static func itemExists(text: String) -> Bool {
+        if let count = RecentSearch.count(predicate: NSPredicate(format: "text == %@", text)), count > 0 {
+            return true
+        }
+        return false
+    }
+    
+    public static func removeItem(text: String) {
+        RecentSearch.deleteAll(predicate: NSPredicate(format: "text == %@", text), context: .new(inMemory: false), includesPropertyValues: false)
+    }
+    
+    public static func removeAll() {
+        RecentSearch.deleteAll()
+    }
+    
+    public static func totalCount() -> Int {
+        let request = getFetchRequest()
+        
+        do {
+            return try DataController.viewContext.count(for: request)
+        } catch {
+            log.error("Count error: \(error)")
+        }
+        return 0
+    }
+    
+    // MARK: - Internal
+    private static func reorderItems(context: NSManagedObjectContext) {
+        DataController.perform(context: .existing(context), save: true) { context in
+            let request = NSFetchRequest<RecentSearch>()
+            request.entity = RecentSearch.entity(context)
+            request.fetchBatchSize = 5
+            
+            let orderSort = NSSortDescriptor(key: "order", ascending: true)
```

Comment:
I don't get it, when you create new recent search you set order to -9999, then you reorder it.
Shouldn't this `order` be updated somewhere?
What is this order for, can't we use `dateAdded` for sorting?

Reply:
Setting the order to `-9999` guarantees the item is at the top of the list. Then sorting them by index and setting the order re-arranges the items. So the item would always be at the top of the list. It is done like this because if someone decides to re-order the search items, then date would be an invalid sorting argument. Similar to playlist, you can set a specific order of an item regardless of date added.

I guess since there is no such requirement here, I can remove it and sort by date, but that was the thinking.